notdiamond/toolkit/rag/evaluation.py

```diff
@@ -0,0 +1,238 @@
+
+    Schema for test_queries can be found below
+    https://docs.ragas.io/en/stable/references/evaluation_schema/#ragas.dataset_schema.SingleTurnSample
+    """
+    samples = []
+    for _, row in test_queries.iterrows():
+        query = row["user_input"]
+        reference = row["reference"]
+        generation_prompt = generation_prompt or row.get("generation_prompt")
+        generator_llm = generator_llm or row.get("generator_llm")
+
+        retrieved_contexts = workflow.get_retrieved_context(query)
+        response = workflow.get_response(query)
+
+        sample = RAGSample(
+            user_input=query,
+            retrieved_contexts=retrieved_contexts,
+            response=response,
+            reference=reference,
+            generation_prompt=generation_prompt,
+            generator_llm=generator_llm,
+        )
+        samples.append(sample)
+    eval_dataset = RAGEvaluationDataset(samples)
+    return eval_dataset
+
+
+def auto_optimize(workflow: BaseNDRagWorkflow, n_trials: int):
+    direction = "maximize" if workflow.objective_maximize else "minimize"
+    study = optuna.create_study(
+        study_name=workflow.job_name, direction=direction
+    )
+    study.optimize(workflow._outer_objective, n_trials=n_trials)
+    workflow._set_param_values(study.best_params)
+    return {"best_params": study.best_params, "trials": study.trials}
+
+
+def _map_to_ragas_samples(
+    dataset: RAGEvaluationDataset,
+) -> Tuple[EvaluationDataset, pd.DataFrame]:
+    ragas_samples = []
+    extra_columns = {
+        "generation_prompt": [],
+    }
+    for sample in dataset:
+        ragas_sample = SingleTurnSample(
+            user_input=sample.user_input,
+            retrieved_contexts=sample.retrieved_contexts,
+            reference_contexts=sample.reference_contexts,
+            response=sample.response,
+            multi_responses=sample.multi_responses,
+            reference=sample.reference,
+            rubrics=sample.rubrics,
+        )
+        ragas_samples.append(ragas_sample)
+        extra_columns["generation_prompt"].append(sample.generation_prompt)
+
+    extra_columns_df = pd.DataFrame.from_dict(extra_columns)
+    return EvaluationDataset(ragas_samples), extra_columns_df
+
+
+def _evaluate_dataset(
+    generator_llm: LLMConfig,
+    dataset: EvaluationDataset,
+    metrics: Optional[Sequence[Metric]] = None,
+    llm: Optional[Union[BaseRagasLLM, LangchainLLM]] = None,
+    embeddings: Optional[
+        Union[BaseRagasEmbeddings, LangchainEmbeddings]
+    ] = None,
+    callbacks: Callbacks = None,
+    in_ci: bool = False,
+    run_config: RunConfig = RunConfig(),
+    token_usage_parser: Optional[TokenUsageParser] = None,
+    raise_exceptions: bool = False,
+    column_map: Optional[Dict[str, str]] = None,
+    show_progress: bool = True,
+    batch_size: Optional[int] = None,
+) -> pd.DataFrame:
+    print(f"Evaluating generations from {str(generator_llm)}")
```

Comment:
Let's double-check for instances of `print` and replace with logger statements as time allows.

Reply:
Done